userguide/02. Intro to Artifacts.ipynb

```diff
@@ -0,0 +1,404 @@
+    "print_chans(chan)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here again we see logging, but a bit more.  Before the deriver is called, flowz first makes sure that each one of the sources is resolved.  Then the deriver is called -- synchronously -- and the results are ready.\n",
+    "\n",
+    "Note above that there are two \"ready\" messages before the first value is actually printed.  That, again, is an indicator of the asychronous processing of the channels.\n",
+    "\n",
+    "In practice, `DerivedArtifacts` are used to gather and transform data that began with `ExtantArtifacts`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "DerivedArtifact starting get\n",
+      "DerivedArtifact waiting on sources.\n",
+      "GuideExtantArtifact<GuideExtantArtifact> starting get\n",
+      "GuideExtantArtifact<GuideExtantArtifact> retrieved.\n",
+      "DerivedArtifact running deriver.\n",
+      "DerivedArtifact ready.\n",
+      "DerivedArtifact starting get\n",
+      "DerivedArtifact waiting on sources.\n",
+      "GuideExtantArtifact<GuideExtantArtifact> starting get\n",
+      "GuideExtantArtifact<GuideExtantArtifact> retrieved.\n",
+      "DerivedArtifact running deriver.\n",
+      "DerivedArtifact ready.\n",
+      "(0, 0)\n",
+      "DerivedArtifact starting get\n",
+      "DerivedArtifact waiting on sources.\n",
+      "GuideExtantArtifact<GuideExtantArtifact> starting get\n",
+      "GuideExtantArtifact<GuideExtantArtifact> retrieved.\n",
+      "DerivedArtifact running deriver.\n",
+      "DerivedArtifact ready.\n",
+      "(1, -10)\n",
+      "(2, -20)\n"
+     ]
+    }
+   ],
+   "source": [
+    "def not_as_lame_deriver(val):\n",
+    "    num, extant_val = val\n",
+    "    return (num, extant_val / -10)\n",
+    "\n",
+    "chan = IterChannel(GuideExtantArtifact(i) for i in range(3)).map(lambda a: DerivedArtifact(not_as_lame_deriver, a))\n",
+    "print_chans(chan)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now you can see that the firing of each `DerivedArtifact` caused its source to be resolved, which meant that its wrapped `GuideExtantArtifact` retrieved its value.  Only after that was the result (a tuple) passed into the deriver."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## ThreadedDerivedArtifact"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A `ThreadedDerivedArtifact` is just like a `DerivedArtifact`, but it is passed a `concurrent.futures.ThreadPoolExecutor` on which it will run.  That can be useful to do perform deriver functions that take a long time (more than X seconds), which might otherwise cause the tornado loop to timeout."
```

Comment:
A few other details are important to convey for this bad boy:

1. Some things are IO-bound and thus quite amenable to the async IO pattern around which tornado is built.  But they _aren't_ implemented in terms of async IO.  In such cases, you can get good results by pushing the blocking IO onto a thread pool executor, which this enables.  The individual threads can block on the synchronous IO, but the main IOLoop continues on its merry way all the while.

   So if you're dealing with synchronous IO-bound clients, put 'em in here.

2. Some routines are pretty hoggy in terms of computation and they'll starve the IOLoop unless you take steps to offload them.  In such cases, getting them onto a thread pool executor (and a shallow pool, at that) can be helpful.

  You're addressing this case but I think it's worth making the distinction between the two.